src/observer.jl

```diff
@@ -1,3 +1,322 @@
+"""
+CircuitObserver is an implementation of an
+observer object (<:AbstractObserver) which
+implements custom measurements to perform
+at each layer of the circuit evolution.
+"""
+struct CircuitObserver <: AbstractObserver
+  observables::Dict{String, Any}
+  functions::Dict{String, Function}
```

Comment:
I was thinking that we could just combine the `functions` and `observables` fields into a single `measurements` field of type `Dict{String, Any}` (or something more specific, like `Dict{String, Union{Function, String, Tuple}}`). Then, we can just check the types of the values stored in the dictionaries, and if they are `String` or `Tuple` treat them as an observable, and if it is a `Function` then call the function.

Reply:
It would make the construction simpler, since then you could just do: `obs = CircuitObserver(["X" => "X", "Y(1)" => ("Y", 1), "norm" => norm])`.